contractcourt/htlc_success_resolver.go

```diff
@@ -201,50 +238,168 @@ func (h *htlcSuccessResolver) Resolve() (ContractResolver, error) {
+		}
 
-	// We'll now broadcast the second layer transaction so we can kick off
-	// the claiming process.
-	//
-	// TODO(roasbeef): after changing sighashes send to tx bundler
-	label := labels.MakeLabel(
-		labels.LabelTypeChannelClose, &h.ShortChanID,
-	)
-	err := h.PublishTx(h.htlcResolution.SignedSuccessTx, label)
-	if err != nil {
-		return nil, err
-	}
+		// Wait for the second level transaction to confirm.
+		// TODO(halseth): could use sweeper result?
+		spendNtfn, err := h.Notifier.RegisterSpendNtfn(
+			&h.htlcResolution.SignedSuccessTx.TxIn[0].PreviousOutPoint,
+			h.htlcResolution.SignDetails.SignDesc.Output.PkScript,
+			h.broadcastHeight,
+		)
+		if err != nil {
+			return nil, err
+		}
 
-	// Otherwise, this is an output on our commitment transaction. In this
-	// case, we'll send it to the incubator, but only if we haven't already
-	// done so.
-	if !h.outputIncubating {
-		log.Infof("%T(%x): incubating incoming htlc output",
+		log.Infof("%T(%x): waiting for second-level HTLC to confirm",
 			h, h.htlc.RHash[:])
 
-		err := h.IncubateOutputs(
-			h.ChanPoint, nil, &h.htlcResolution,
-			h.broadcastHeight,
+		var (
+			spend *chainntnfs.SpendDetail
+			ok    bool
+		)
+		select {
+		case spend, ok = <-spendNtfn.Spend:
+			if !ok {
+				return nil, errResolverShuttingDown
+			}
+
+		case <-h.quit:
+			return nil, errResolverShuttingDown
+		}
+
+		// Now that the second level transaction has confirmed, we
+		// report the resolver has moved the next stage.
+		h.reportLock.Lock()
+		h.currentReport.Stage = 2
+		h.reportLock.Unlock()
+
+		waitHeight := uint32(spend.SpendingHeight) +
+			h.htlcResolution.CsvDelay - 1
+		log.Infof("%T(%x): waiting for CSV lock to expire at height %v",
+			h, h.htlc.RHash[:], waitHeight)
+
+		if err := waitForHeight(waitHeight, h.Notifier, h.quit); err != nil {
+			return nil, err
+		}
+
+		// Find output index of the second-level, as the sweeper might
+		// have re-arranged outputs.
+		txOut := h.htlcResolution.SignedSuccessTx.TxOut[0]
+		index := -1
+		for i, o := range spend.SpendingTx.TxOut {
+			if bytes.Equal(o.PkScript, txOut.PkScript) {
+				index = i
+			}
+		}
+
+		if index == -1 {
+			err := fmt.Errorf("unable to find second-level "+
+				"output %v on spending tx %v", txOut,
+				spend.SpenderTxHash)
+			log.Error(err)
+			return nil, err
+		}
+
+		op := &wire.OutPoint{
+			Hash:  *spend.SpenderTxHash,
+			Index: uint32(index),
+		}
+
+		log.Infof("%T(%x): CSV lock expired, offering second-layer "+
+			"output to sweeper: %v", h, h.htlc.RHash[:], op)
+
+		inp := input.NewCsvInput(
+			op, input.HtlcAcceptedSuccessSecondLevel,
+			&h.htlcResolution.SweepSignDesc, h.broadcastHeight,
+			h.htlcResolution.CsvDelay,
+		)
+		_, err = h.Sweeper.SweepInput(
+			inp,
+			sweep.Params{
+				Fee: sweep.FeePreference{
+					ConfTarget: sweepConfTarget,
+				},
+			},
 		)
 		if err != nil {
 			return nil, err
 		}
 
-		h.outputIncubating = true
+		// TODO(halseth): need to checkpoint?
```

Comment:
Yes, as otherwise, it'll try to re offer the input to the sweeper, which will error out as the output has already been spent (after the max sweep attempts). Alternatively, you can move things around a bit into two functions, and enter them based on spend/conf notification.

Reply:
Now checkpointing is added :)